README.md
![tecandidates_logos](https://user-images.githubusercontent.com/31257257/29850516-b42a2540-8d04-11e7-9286-7e8ebefbd455.png)

# TEcandidates
TEcandidates is a pipeline to include transposable elements in RNA-seq differential expression analysis.


### INSTALLATION INSTRUCTIONS
_________________________________________________

#### 1. Dependencies

TEcandidates is implemented in Bash, and requires no installation. However, other softwares that are part of the pipeline are required. The following are the required softwares, along with some minimum instructions to install them in a computer with Linux. For additional help and/or troubleshooting, the source page of each software is listed for more detailed instructions.

- **BEDtools v2.25**

-Source page: https://github.com/arq5x/bedtools2/releases/tag/v2.25.0

    $ wget https://github.com/arq5x/bedtools2/releases/download/v2.25.0/bedtools-2.25.0.tar.gz
    $ tar -xvzf bedtools-2.25.0.tar.gz
    $ cd bedtools2/
    $ make
    $ sudo make install

Check if installation was correct using        

    $ bedtools --version
    bedtools v2.25.0

- **BioPerl**

-Source page: http://bioperl.org/INSTALL.html

	$ sudo perl -e shell -MCPAN
	Password:
	Terminal does not support AddHistory.

	cpan shell -- CPAN exploration and modules installation (v1.9800)
	Enter 'h' for help.

	cpan[1]> install C/CJ/CJFIELDS/BioPerl-1.007001.tar.gz

Verify installation with:

	$ perl -MBio::Root::Version -e 'print $Bio::Root::Version::VERSION,"\n"'

        
- **Bowtie v2.3**

-Source page: https://sourceforge.net/projects/bowtie-bio/files/bowtie2/2.3.2

Download the appropiate file for your distribution and follow these instructions:

    $ unzip bowtie2-2.3.2-linux-x86_64.zip
    $ cd bowtie2-2.3.2/
    $ pwd

The above will show the full path to the bowtie2 binaries. Copy it and add it to the $PATH environment variable:

    $ export PATH=$PATH:path_to_bowtie2-2.3.2

Check correct installation with

    $ bowtie2 --version
    bowtie2-align-s version 2.3.2
    64-bit
    Built on dde45b53bd81
    Sat May  6 02:39:52 UTC 2017
    Compiler: gcc version 4.1.2 20080704 (Red Hat 4.1.2-55)
    Options: -O3 -m64 -msse2 -funroll-loops -g3 -DPOPCNT_CAPABILITY -DWITH_TBB -DNO_SPINLOCK -DWITH_QUEUELOCK=1
    Sizeof {int, long, long long, void*, size_t, off_t}: {4, 8, 8, 8, 8, 8}


- **Samtools v1.4.1**

-Source page: https://github.com/samtools/samtools

	$ wget https://github.com/samtools/samtools/releases/download/1.4.1/samtools-1.4.1.tar.bz2
	$ tar -xvjf samtools-1.4.1.tar.bz2
	$ cd samtools-1.4.1
	$ ./configure
	$ make
	$ sudo make install

Check correct installation:
	
	$ samtools --version
	samtools 1.4.1
	Using htslib 1.4.1
	Copyright (C) 2017 Genome Research Ltd.

- **Trinity v2.4**

-Source page: https://github.com/trinityrnaseq/trinityrnaseq/

    $ wget https://github.com/trinityrnaseq/trinityrnaseq/archive/master.zip
    $ unzip master.zip
    $ cd trinityrnaseq-master/
    $ make
    $ pwd 

Copy the path to the Trinity directory and add it to the $PATH environment variable:

    $ export PATH=$PATH:path_to_Trinity

Check correct installation with

    $ Trinity --version
    Trinity version: Trinity-v2.4.0

#### 2. Installing the TEcandidates pipeline

Download the TEcandidates tarball, and uncompress it:

	$ wget https://github.com/TEcandidates/TEcandidates/blob/master/TEcandidates_LATEST_STABLE.tar.gz
    $ tar -xvzf TEcandidates_LATEST_STABLE.tar.gz

Grant execution permissions to the pipeline script:

    $ chmod u+x TEcandidates_LATEST_STABLE/TEcandidates.sh

For simplicity of use, add the TEcandidates full path to your PATH environment variable. First get the full path:

    $ readlink -f TEcandidates_LATEST_STABLE
    
Then copy the output of the previous command, and add it to the PATH variable:

    $ export PATH=$PATH:/path/to/TEcandidates_LATEST_STABLE

### SAMPLE USAGE
_________________________________________________

Once TEcandidates is in your PATH variable, you can execute it as

    $ TEcandidates.sh -t=Number_of_threads -r=RAM_to_use -g=Genome_Fasta_File -fq=Path_to_FASTQ_files -m=Mode -te=TE_Annotation -c=Coverage -l=TE_length -N=Number_of_candidateTEs

    -t Number of threads to use in the softwares executed during the pipeline
    -r Maximum amount of RAM assigned to Trinity (Trinity's --max_memory option)
    -g Genome to use (FASTA format, .fasta extension)
    -fq Path to FASTQ files to use (all files must have .fastq extension)
    -m Mode of FASTQ files, SE for Single-end reads and PE for Paired-end reads
    -te Transposable Element annotation file
    -c Minimum coverage in which a Transposable Element must be covered by a de-novo transcript in order to be selected as candidate
    -l Minimum length of Transposable Element to be considered in the selection step
    -N Number of candidate Transposable Elements to output


**Important considerations**
- Reads files must have ".fastq" extension.
- TEcandidates can be used with either single-end reads or paired-end reads.  
Paired-end reads **must have** "\_1.fastq" and "\_2.fastq" extensions.

### SAMPLE DATA
_________________________________________________

In order to check that TEcandidates is working correctly, please test it with a dataset from _Drosophila melanogaster_ (Ohtani et al., 2013). The dataset is available at Gene Expression Omnibus, accession no. GSE47006, and must be downloaded with The _fastq-dump_ tool from the SRA toolkit . To install the SRA toolkit, please copy the link of the appropriate version for your Operating System from https://www.ncbi.nlm.nih.gov/sra/docs/toolkitsoft/, and download it:

    $ wget https://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/current/sratoolkit.current-ubuntu64.tar.gz

Check if fastq-dump is working:

    $ tar -xvzf sratoolkit.current-ubuntu64.tar.gz
    $ ./sratoolkit.2.8.2-1-ubuntu64/bin/fastq-dump -V

      ./sratoolkit.2.8.2-1-ubuntu64/bin/fastq-dump : 2.8.2

For ease of use, add the SRA toolkit to your $PATH environment variable. Get the full path to the SRA toolkit binaries with 

    $ readlink -f sratoolkit.2.8.2-1-ubuntu64/bin/

and append it to $PATH like this:

    $ export PATH=$PATH:path_to_SRAtoolkit


Create a new work directory

    $ mkdir TEcandidates_test
    $ cd TEcandidates_test

Download the _Drosophila melanogaster_ control dataset with:

    $ nohup fastq-dump --defline-seq '@$sn[_$rn]/$ri' --split-files --accession SRR851837 > SRR851837.fastq-dump.log &

Download the _Drosophila melanogaster_ treatment dataset with:

    $ nohup fastq-dump --defline-seq '@$sn[_$rn]/$ri' --split-files --accession SRR851838 > SRR851838.fastq-dump.log &
    
Once these processes are done, check that they were downloaded correctly with 

    $ tail *log
    ==> SRR851837.fastq-dump.log <==
    Read 42134407 spots for SRR851837
    Written 42134407 spots for SRR851837

    ==> SRR851838.fastq-dump.log <==
    Read 48277060 spots for SRR851838
    Written 48277060 spots for SRR851838

Download the genome and the annotation file:

	$ wget http://mobilomics.cl/tecandidates/files/dm3.fasta
	$ wget http://mobilomics.cl/tecandidates/files/dm3_rmsk_TE.gff3

Execute the pipeline script afterwards:

    nohup TEcandidates.sh -t=10 -r=128 -c=0.3 -l=900 -te=dm3_rmsk_TE.gff3 -g=dm3.fasta -fq=. -m=SE -N=1 > TEcandidates.log &


Once it's done, you should have a folder named **candidateTE_analysis_coverage-0.3_length-900_N-1**, that contains the following files:

    $ ls -lht candidateTE_analysis_coverage-0.3_length-900_N-1
	total 2.1G
	drwxr-xr-x 2 bvaldebenito bvaldebenito 4.0K Apr 28 08:22 trinity_assemblies
	-rw-r--r-- 1 bvaldebenito bvaldebenito  44M Apr 28 08:23 dm3.fasta.masked_BT2.rev.1.bt2
	-rw-r--r-- 1 bvaldebenito bvaldebenito  30M Apr 28 08:23 dm3.fasta.masked_BT2.rev.2.bt2
	-rw-r--r-- 1 bvaldebenito bvaldebenito  44M Apr 28 08:22 dm3.fasta.masked_BT2.1.bt2
	-rw-r--r-- 1 bvaldebenito bvaldebenito  30M Apr 28 08:22 dm3.fasta.masked_BT2.2.bt2
	-rw-r--r-- 1 bvaldebenito bvaldebenito 505K Apr 28 08:22 dm3.fasta.masked_BT2.3.bt2
	-rw-r--r-- 1 bvaldebenito bvaldebenito  30M Apr 28 08:22 dm3.fasta.masked_BT2.4.bt2
	-rw-r--r-- 1 bvaldebenito bvaldebenito 165M Apr 28 08:22 dm3.fasta.masked
	-rw-r--r-- 1 bvaldebenito bvaldebenito 5.8M Apr 28 08:22 repeatsToMask_coverage-0.3_length-900.gff3
	-rw-r--r-- 1 bvaldebenito bvaldebenito 7.4K Apr 28 08:22 allcandidates_coverage-0.3_length-900_N-1.gff3
	-rw-r--r-- 1 bvaldebenito bvaldebenito 1.1G Apr 28 08:01 SRR851838_1_filtered.fastq
	-rw-r--r-- 1 bvaldebenito bvaldebenito  223 Apr 28 07:57 SRR851838_1.bt2_summary
	-rw-r--r-- 1 bvaldebenito bvaldebenito 718M Apr 28 07:39 SRR851837_1_filtered.fastq
	-rw-r--r-- 1 bvaldebenito bvaldebenito  222 Apr 28 07:36 SRR851837_1.bt2_summary


The candidate Transposable Elements can be found in the **allcandidates_coverage-0.3_length-900_N-1.gff3**, those that were removed from the genome in the **repeatsToMask_coverage-0.3_length-900.gff3** file. The genome, with the repeats in **repeatsToMask_coverage-0.3_length-900.gff3** removed, is in the **dm3.fasta.masked**. Additional files for Bowtie 2 are also generated (**\*\_BT2\*.bt2**).

### CONTACT
_________________________________________________

Please send any inquiries about usage and/or bugs to TEcandidates@gmail.com




### REFERENCES
_________________________________________________

	Ohtani H, Iwasaki YW, Shibuya A, Siomi H, Siomi MC, Saito K. (2013). DmGTSF1 is necessary for Piwi-piRISC-mediated transcriptional transposon silencing in the Drosophila ovary. Genes Dev. 2013 Aug 1;27(15):1656-61. doi: 10.1101/gad.221515.113.





TEcandidates_v2.0.1/CHANGELOG.txt
v2.0.1
April 17th,2018
- PLACEHOLDER FOR PE READS 

April 13th, 2018
- Added pre-mapping step to diminish the number of reads to be used during assembly with Trinity
- Added option to output any number of candidates (DEFAULT N = 1, this was also the default for all previous versions).

v2.0
March 13th, 2018
- Fixed some issues with Trinity Butterfly that caused the pipeline to use up all the memory
- Added the option to provide a custom length for the candidate selection step
- Added a util that allows users to use a previous run of the pipeline to generate results with another length and/or coverage

TO-DO
- Fix issue that after failing Trinity, keeps running the pipeline (should be a validation for Trinity successful assembly)


TEcandidates_LATEST_STABLE/utils/rerun.sh
#!/bin/bash

for i in "$@"
do
	case $i in
	-g=*|--genome=*)
	Genome="${i#*=}"
	;;
	-t=*|--threads=*)
	threads="${i#*=}"
	;;
        -c=*|--coverage=*)
        minimumCoverage="${i#*=}"
        ;;
	-l=*|--length=*)
        minimumLength="${i#*=}"
        ;;
	-te=*|--te-annotation=*)
        TEannotation="${i#*=}"
        ;;
	-r=*|--RAM=*)
        RAM="${i#*=}"
        ;;
	-d=*|--directory=*)
        previousRunDirectory="${i#*=}"
        ;;
	esac
done

####DEPENDENCIES VALIDATION


echo "Checking if Bedtools v2.25 is installed.."
if hash bedtools 2>/dev/null; then
        bedtools_version=$(bedtools --version)
        if [[ $bedtools_version == "bedtools v2.25.0" ]]; then
                echo $bedtools_version
                echo "Bedtools v2.25 is installed!"
        fi
else
        echo "Bedtools v2.25 was not found. Please install it and try again."
        exit
fi
echo ""


echo "Checking if Bioperl is installed.."
bioperl_version=$(perl -MBio::Root::Version -e 'print $Bio::Root::Version::VERSION,"\n"')
if [[ $bioperl_version != *"Can't locate"* ]]; then
        echo $bioperl_version
        echo "Bioperl is installed!"
else
        echo "Bioperl was not found. Please install it and try again."
        exit
fi
echo ""


echo "Checking if Bowtie v2.3 is installed.."
if hash bowtie2 2>/dev/null; then
        bowtie2_version=$(bowtie2 --version|head -n 1)
        if [[ $bowtie2_version == *"bowtie2-align-s version 2.3.0"* ]]; then
                echo $bowtie2_version
                echo "Bowtie v2.3 is installed!"
        fi
else
        echo "Bowtie2 v2.3 was not found. Please install it and try again."
        exit
fi
echo ""


####INPUT VALIDATION

if [ -z $Genome ]
then
	echo "Please specify a Genome FASTA file"
	exit
fi

if [ -z $threads ]
then
        echo "Please specify a number of threads to use during the pipeline execution"
        exit
fi

if [ -z $RAM ]
then
        echo "Please specify the amount of RAM (in GB) to use during the pipeline execution"
        exit
fi

if [ -z $previousRunDirectory ]
then
        echo "Please specify the folder containing the previous run results"
        exit
fi

if [ -z $minimumCoverage ]
then
        echo "Please specify the coverage at which candidate TEs will be selected"
        exit
fi

if [ -z $minimumLength ]
then
        echo "Please specify the length at which candidate TEs will be selected"
        exit
fi

if [ -z $TEannotation ]
then
        echo "Please specify a TE annotation file in GFF3 format"
        exit
fi


CPU=$threads
bflyHeapMax=$(($RAM/$CPU))
bflyHeapMax=$bflyHeapMax"G"
RAM=$RAM"G"
previousRunDirectory=$(readlink -f $previousRunDirectory)

####PIPELINE START

Genome=$(readlink -f $Genome)
TEannotation=$(readlink -f $TEannotation)

echo "Genome file: $Genome"
echo "Transposable element annotation: $TEannotation"
echo "Coverage: $minimumCoverage"
echo "Memory: $RAM"
echo "Threads: $threads"
echo "Previous run directory: $previousRunDirectory"

CPU=$threads

currentworkingdir=$(pwd)

TEcandidatesDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"

cd $currentworkingdir

outputdir="candidateTE_analysis_coverage-"$minimumCoverage"_length-"$minimumLength

if [ ! -d $outputdir ];
then
	echo "Creating $outputdir ..."
	mkdir $outputdir
else
	echo "Output directory already exists"
fi

outputdir=$(readlink -f $outputdir)
cd $outputdir
mkdir trinity_assemblies

echo "#####GENERATING CANDIDATES"

for trinityBed in $previousRunDirectory/trinity_assemblies/*.trinity_assembly.bed;
do
        echo "Generating candidate TE file for $trinityBed..."
	
        candidateTEfilename=${trinityBed/.trinity_assembly.bed/.candidate_TEs}
	candidateTEfilename=$(basename $candidateTEfilename)
	candidateTEfilename=$outputdir"/trinity_assemblies/"$candidateTEfilename	

	
	cmd="bedtools coverage -s -a $TEannotation -b $trinityBed |awk -v minimumCoverage=$minimumCoverage -v minimumLength=$minimumLength '(\$(NF-1)>=minimumLength && \$NF>=minimumCoverage) {print \$0}' > $candidateTEfilename"
	echo -e "CMD: $cmd\n"
	bedtools coverage -s -a $TEannotation -b $trinityBed |awk -v minimumCoverage=$minimumCoverage -v minimumLength=$minimumLength '($(NF-1)>=minimumLength && $NF>=minimumCoverage) {print $0}' > $candidateTEfilename

	echo -e "$candidateTEfilename generated!\n\n"
done

cd "trinity_assemblies"
cat *.candidate_TEs > allcandidates.gff3

#Sort the file with the coverage info, according to their ID (column 9), then by the TE instance length (column 12) and by the coverage by a de novo contig (column 13). 
sort -k9,9d -k12,12n -k13,13n allcandidates.gff3 > allcandidates.gff3.sorted
finalfile="allcandidates_coverage-"$minimumCoverage"_length-"$minimumLength".gff3"

echo $finalfile

#Column 9 is where the ID of a TE is. Here, the files with the coverages per TE instance is grouped by their ID, in order to find the instance of a given TE that has the greates coverage. Then, the location is appended to the ID (this might be redundant, as this is a GFF3 file after all).
bedtools groupby -g 9 -c 1,2,3,4,5,6,7,8,9 -o last -i allcandidates.gff3.sorted|awk 'BEGIN{OFS="\t"} {$10=$10"/ChrID="$2$8; print $2,$3,$4,$5,$6,$7,$8,$9,$10}' > $finalfile

mv $finalfile ../
cd ..

echo "#####GENERATING MASKED GENOME AND FINAL FILES"
repeatsToMask="repeatsToMask_coverage-"$minimumCoverage"_length-"$minimumLength".gff3"
cmd="bedtools intersect -v -a $TEannotation -b $finalfile > $repeatsToMask"
echo -e "CMD: $cmd\n"
bedtools intersect -v -a $TEannotation -b $finalfile > $repeatsToMask

maskedGenome=$(basename $Genome)
maskedGenome=$maskedGenome".masked"
BT2_basename=$maskedGenome"_BT2"
cmd="bedtools maskfasta -fi $Genome -fo $maskedGenome -bed $repeatsToMask -mc X"
echo -e "CMD: $cmd\n"
bedtools maskfasta -fi $Genome -fo $maskedGenome -bed $repeatsToMask -mc X

cmd="bowtie2-build --threads $threads $maskedGenome $BT2_basename"
echo -e "CMD: $cmd\n"
bowtie2-build --threads $threads $maskedGenome $BT2_basename



TEcandidates_v2.0.1/TEcandidates.sh
#!/bin/bash

for i in "$@"
do
	case $i in
	-g=*|--genome=*)
	Genome="${i#*=}"
	;;
	-t=*|--threads=*)
	threads="${i#*=}"
	;;
	-m=*|--mode=*)
        mode="${i#*=}"
        ;;
	-fq=*|--fastq-files=*)
	fastqFiles="${i#*=}"
	;;
        -c=*|--coverage=*)
        minimumCoverage="${i#*=}"
        ;;
	-l=*|--length=*)
        minimumLength="${i#*=}"
        ;;
	-te=*|--te-annotation=*)
        TEannotation="${i#*=}"
        ;;
	-r=*|--RAM=*)
        RAM="${i#*=}"
        ;;
	-N=*)
        N="${i#*=}"
        ;;
	esac
done

####DEPENDENCIES VALIDATION


echo "Checking if Bedtools v2.25 is installed.."
if hash bedtools 2>/dev/null; then
        bedtools_version=$(bedtools --version)
        if [[ $bedtools_version == "bedtools v2.25.0" ]]; then
                echo $bedtools_version
                echo "Bedtools v2.25 is installed!"
        fi
else
        echo "Bedtools v2.25 was not found. Please install it and try again."
        exit
fi
echo ""


echo "Checking if Bioperl is installed.."
bioperl_version=$(perl -MBio::Root::Version -e 'print $Bio::Root::Version::VERSION,"\n"')
if [[ $bioperl_version != *"Can't locate"* ]]; then
        echo $bioperl_version
        echo "Bioperl is installed!"
else
        echo "Bioperl was not found. Please install it and try again."
        exit
fi
echo ""


echo "Checking if Bowtie v2.3 is installed.."
if hash bowtie2 2>/dev/null; then
        bowtie2_version=$(bowtie2 --version|head -n 1)
        if [[ $bowtie2_version == *"bowtie2-align-s version 2.3.0"* ]]; then
                echo $bowtie2_version
                echo "Bowtie v2.3 is installed!"
        fi
else
        echo "Bowtie2 v2.3 was not found. Please install it and try again."
        exit
fi
echo ""


####INPUT VALIDATION

if [ -z $Genome ]
then
	echo "Please specify a Genome FASTA file"
	exit
fi

if [ -z $threads ]
then
        echo "Please specify a number of threads to use during the pipeline execution"
        exit
fi

if [ -z $RAM ]
then
        echo "Please specify the amount of RAM (in GB) to use during the pipeline execution"
        exit
fi

if [ -z $mode ]
then
        echo "Please specify a mode: SE for Single-End Reads or PE for Paired-End Reads"
        exit
fi

if [ -z $fastqFiles ]
then
        echo "Please specify the folder containing the FASTQ files"
        exit
fi

if [ -z $minimumCoverage ]
then
        echo "Please specify the coverage at which candidate TEs will be selected"
        exit
fi

if [ -z $minimumLength ]
then
        echo "Please specify the length at which candidate TEs will be selected"
        exit
fi

if [ -z $TEannotation ]
then
        echo "Please specify a TE annotation file in GFF3 format"
        exit
fi


CPU=$threads
bflyHeapMax=$(($RAM/$CPU))
bflyHeapMax=$bflyHeapMax"G"
RAM=$RAM"G"
####PIPELINE START


if [ "$mode" == "SE" ];
then
	echo "TEcandidates started in Single-End mode"
	fastqFiles=$fastqFiles"/*fastq"
else
	echo "TEcandidates started in Paired-End mode"
	fastqFiles=$fastqFiles"/*_1.fastq"
fi

readFiles=$(readlink -f $fastqFiles)
echo "The following read files will be used for the pipeline:"
echo $readFiles

Genome=$(readlink -f $Genome)
TEannotation=$(readlink -f $TEannotation)

echo "Genome file: $Genome"
echo "Transposable element annotation: $TEannotation"
echo "Coverage: $minimumCoverage"
echo "Memory: $RAM"
echo "Threads: $threads"
echo "Number of candidates per TE instance: $N"

CPU=$threads

currentworkingdir=$(pwd)

TEcandidatesDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"

cd $currentworkingdir

outputdir="candidateTE_analysis_coverage-"$minimumCoverage"_length-"$minimumLength"_N-"$N

if [ ! -d $outputdir ];
then
	echo "Creating $outputdir ..."
	mkdir $outputdir
else
	echo "Output directory already exists"
fi

cd $outputdir



echo "#####PRE-MAPPING READS"

BT2_basename=${Genome/.fasta/_BT2}

echo "Building BT2 index for original genome"
bowtie2-build $Genome $BT2_basename -q --threads $threads
echo "Done building BT2 index"

filteredReadFiles=""

for readFile in $readFiles;
do

	readFileBasename=$(basename $readFile)

	if [ "$mode" == "SE" ];
        then
		readFileBasename=$(basename $readFile)
		samFile=${readFileBasename/.fastq/.sam}
		bowtie2Summary=${readFileBasename/.fastq/.bt2_summary}
		bamFile=${samFile/.sam/.bam}
		filteredBamFile=${samFile/.sam/_filtered.bam}
		filteredFastq=${filteredBamFile/.bam/.fastq}

		if [ -e $filteredFastq ];
		then
			echo "Filtered Fastq for $readFile exists! Filename: $filteredFastq"
		else
			echo "Mapping reads from $readFile to genome"
			bowtie2 -N 1 --threads $threads -x $BT2_basename -U $readFile -S $samFile 2> $bowtie2Summary
			echo "Pre-mapping finished"

			echo "Creating filtered reads file"
			#samtools view $samFile --threads $threads -b > $bamFile
			#bedtools intersect -a $bamFile -b $TEannotation -u > $filteredBamFile
			samtools view $samFile --threads $threads -b|bedtools intersect -a - -b $TEannotation -u > $filteredBamFile
			samtools fastq $filteredBamFile --threads $threads > $filteredFastq

			#Conversion to FASTA input (if required)
			#awk '(NR%4==1||NR%4==2){print $0}' $filteredFastq|sed 's/^@/>/g' > temp.txt
			#mv temp.txt $filteredFastq

			rm $samFile $filteredBamFile
			echo "Done creating filtered reads file $filteredFastq"
		fi

		filteredFastq=$(readlink -f $filteredFastq)
		filteredReadFiles=$filteredReadFiles""$filteredFastq" "
        else
        	#Placeholder for Paired-end reads
		mate1=$readFileBasename
                mate2=${mate1/_1.fastq/_2.fastq}

		samFile=${readFileBasename/_1.fastq/.sam}		
		bowtie2Summary=${readFileBasename/_1.fastq/.bt2_summary}
		bamFile=${samFile/.sam/.bam}
		filteredBamFile=${samFile/.sam/_filtered.bam}
		filteredFastq=${filteredBamFile/.bam/.fastq}

		if [ -e $filteredFastq ];
                then
			echo "Filtered Fastq for $readFile exists! Filename: $filteredFastq"		
		else
			exit
			#bowtie2 -N 1 --threads $threads -x $BT2_basename -1 $mate1 -2 $mate2 2> $bowtie2Summary|samtools view 	
		

		fi

 
	fi

        
done
######

echo "#####GENERATING TRINITY ASSEMBLIES"

if [ ! -d trinity_assemblies ];
then
	mkdir trinity_assemblies
fi

#for readFile in $readFiles;
for readFile in $filteredReadFiles;
do

	echo "Generating Trinity assembly with $readFile"
	basename=$(basename $readFile)
	timeOutput="trinity_assemblies/"${basename/.fastq/.time}
	trinityOutput="trinity_assemblies/"${basename/.fastq/.trinity_assembly.fasta}

	if [ -e $trinityOutput ];
	then
		echo "$trinityOutput exists. Skipping this file"
	else
		if [ "$mode" == "SE" ];
		then
			cmd="/usr/bin/time -f \"%E real\n%U user\n%S sys\n%K memory\" -o $timeOutput Trinity --seqType fq --max_memory $RAM --CPU $CPU --bflyHeapSpaceMax $bflyHeapMax --bflyCPU $CPU --single $readFile --full_cleanup"
			echo "CMD: $cmd"
			echo -e "\n"
			#Trinity --seqType fq --single $readFile --max_memory $RAM --CPU $CPU --bflyHeapSpaceMax $RAM --bflyCPU $CPU --full_cleanup
			/usr/bin/time -f "%E real\n%U user\n%S sys\n%K memory" -o $timeOutput Trinity --seqType fq --max_memory $RAM --CPU $CPU --bflyHeapSpaceMax $bflyHeapMax --bflyCPU $CPU --single $readFile --full_cleanup
		else
			mate1=$readFile
			mate2=${mate1/_1.fastq/_2.fastq}
			cmd="/usr/bin/time -f \"%E real\n%U user\n%S sys\n%K memory\" -o $timeOutput Trinity --seqType fq --max_memory 128G --CPU 12 --bflyHeapSpaceMax 10G --bflyCPU 12 --left $mate1 --right $mate2 --full_cleanup"
			Trinity --seqType fq --max_memory $RAM --CPU $CPU --bflyHeapSpaceMax $RAM --bflyCPU $CPU --left $mate1 --right $mate2 --full_cleanup
		fi
	
		cmd="mv trinity_out_dir.Trinity.fasta $trinityOutput"
		echo -e "CMD: $cmd\n"
		mv trinity_out_dir.Trinity.fasta $trinityOutput

		if [ $? -ne 0 ];
		then
			"Could not execute $cmd"
			exit
		fi
		echo "Trinity output stored in $trinityOutput"
	fi

	parsed_trinityOutput=$trinityOutput".parsed"
	cmd="perl $TEcandidatesDIR/utils/parseFasta.pl $trinityOutput > $parsed_trinityOutput"
	echo -e "CMD: $cmd\n"
	perl $TEcandidatesDIR/utils/parseFasta.pl $trinityOutput > $parsed_trinityOutput
done

###
echo "#####MAPPING ASSEMBLIES INTO GENOME"

#BT2_basename=$(basename $Genome)
#BT2_basename=${BT2_basename/.fasta/_BT2}


cmd="bowtie2-build --threads $threads $Genome $BT2_basename"
echo -e "CMD: $cmd"
#bowtie2-build --threads $threads $Genome $BT2_basename

for assemblyFile in trinity_assemblies/*.trinity_assembly.fasta.parsed;
do
	echo "Mapping $assemblyFile into $BT2_basename"
	samFile=${assemblyFile/fasta.parsed/sam}
	bowtie2Summary=${assemblyFile/fasta.parsed/bt2_summary}
	cmd="bowtie2 -N 1 --threads $threads -f -x $BT2_basename -U $assemblyFile -S $samFile 2> $bowtie2Summary"
	echo -e "CMD: $cmd\n"

	if [ ! -e $bowtie2Summary ]
	then
		bowtie2 -N 1 --threads $threads -f -x $BT2_basename -U $assemblyFile -S $samFile 2> $bowtie2Summary
	else
		if [ $(grep -c "overall" $bowtie2Summary) -eq 0 ]
		then
			bowtie2 -N 1 --threads $threads -f -x $BT2_basename -U $assemblyFile -S $samFile 2> $bowtie2Summary
		else
			echo "$assemblyFile already mapped. Skipping this step"
		fi	
	fi

	bamFile=${samFile/sam/bam}

	if [ -e $samFile ]
	then
		cmd="samtools view -o $bamFile --threads $threads -b $samFile"
		echo -e "CMD: $cmd\n"
		samtools view -o $bamFile --threads $threads -b $samFile
		rm $samFile
	fi

	bedFile=${samFile/sam/bed}
	cmd="bedtools bamtobed -i $bamFile > $bedFile"
	echo -e "CMD: $cmd\n"
	bedtools bamtobed -i $bamFile > $bedFile

done

cmd="rm *$BT2_basename*"
echo -e "CMD: $cmd\n"
rm *$BT2_basename*
####

echo "#####GENERATING CANDIDATES"

for trinityBed in trinity_assemblies/*.trinity_assembly.bed;
do
        echo "Generating candidate TE file for $trinityBed..."
        candidateTEfilename=${trinityBed/.trinity_assembly.bed/.candidate_TEs}
        ##candidateTEfilename=$outputdir"/"$candidateTEfilename
        #bedtools coverage -s -a $TEannotation -b $trinityBed |awk -v minimumCoverage=$minimumCoverage '($(NF-1)>900 && $NF>=minimumCoverage) {print $0}' > $candidateTEfilename

	cmd="bedtools coverage -s -a $TEannotation -b $trinityBed |awk -v minimumCoverage=$minimumCoverage -v minimumLength=$minimumLength '(\$(NF-1)>=minimumLength && \$NF>=minimumCoverage) {print \$0}' > $candidateTEfilename"
	echo -e "CMD: $cmd\n"
	bedtools coverage -s -a $TEannotation -b $trinityBed |awk -v minimumCoverage=$minimumCoverage -v minimumLength=$minimumLength '($(NF-1)>=minimumLength && $NF>=minimumCoverage) {print $0}' > $candidateTEfilename
	echo -e "$candidateTEfilename generated!\n\n"
done

cd "trinity_assemblies"
cat *.candidate_TEs > allcandidates.gff3
sort -k9,9d -k12,12n -k13,13n allcandidates.gff3 > allcandidates.gff3.sorted
finalfile="allcandidates_coverage-"$minimumCoverage"_length-"$minimumLength"_N-"$N".gff3"

echo $finalfile

#################

allcandidatesFile="allcandidates.gff3.sorted"
idsFile="uniqueIDS.txt"
groupedcandidatesFile=$allcandidatesFile".grouped"

bedtools groupby -g 1,2,3,4,5,6,7,8,9 -c 10,11,12,13 -o last -i $allcandidatesFile > $groupedcandidatesFile

awk '{print $9}' $allcandidatesFile|sort -u > $idsFile
while read line
do
        pattern=$line
        cmd="grep -w $pattern $groupedcandidatesFile|tail -n $N >> $finalfile"
        #echo "$cmd"
        grep -w $pattern $groupedcandidatesFile|tail -n $N >> $finalfile
done < "$idsFile"

#################
#bedtools groupby -g 9 -c 1,2,3,4,5,6,7,8,9 -o last -i allcandidates.gff3.sorted|awk 'BEGIN{OFS="\t"} {$10=$10"/ChrID="$2$8; print $2,$3,$4,$5,$6,$7,$8,$9,$10}' > $finalfile

awk 'BEGIN{OFS="\t"} {$9=$9"/ChrID="$1$7"_"$4"-"$5; print $1,$2,$3,$4,$5,$6,$7,$8,$9}' $finalfile > temp.txt
#awk 'BEGIN{OFS="\t"} {$9=$9"_"$4"-"$5; print $1,$2,$3,$4,$5,$6,$7,$8,$9}' $finalfile > temp.txt
mv temp.txt $finalfile 

mv $finalfile ../
cd ..

echo "#####GENERATING MASKED GENOME AND FINAL FILES"
repeatsToMask="repeatsToMask_coverage-"$minimumCoverage"_length-"$minimumLength".gff3"
cmd="bedtools intersect -v -a $TEannotation -b $finalfile > $repeatsToMask"
echo -e "CMD: $cmd\n"
bedtools intersect -v -a $TEannotation -b $finalfile > $repeatsToMask

maskedGenome=$(basename $Genome)
maskedGenome=$maskedGenome".masked"
BT2_basename=$maskedGenome"_BT2"
cmd="bedtools maskfasta -fi $Genome -fo $maskedGenome -bed $repeatsToMask -mc X"
echo -e "CMD: $cmd\n"
bedtools maskfasta -fi $Genome -fo $maskedGenome -bed $repeatsToMask -mc X

cmd="bowtie2-build --threads $threads $maskedGenome $BT2_basename"
echo -e "CMD: $cmd\n"
bowtie2-build --threads $threads $maskedGenome $BT2_basename

